arbitrator/prover/src/merkle.rs

```diff
@@ -146,44 +355,221 @@ impl Merkle {
+        *self = Self::new_advanced(self.ty, leaves, self.min_depth);
     }
 
     /// Removes the rightmost leaf from the merkle
     /// Currently O(n) in the number of leaves (could be log(n))
     pub fn pop_leaf(&mut self) {
-        let mut leaves = self.layers.swap_remove(0);
+        let mut leaves = self.layers.lock().data.swap_remove(0);
         leaves.pop();
-        let empty = self.empty_layers[0];
-        *self = Self::new_advanced(self.ty, leaves, empty, self.min_depth);
+        *self = Self::new_advanced(self.ty, leaves, self.min_depth);
     }
 
-    pub fn set(&mut self, mut idx: usize, hash: Bytes32) {
-        if self.layers[0][idx] == hash {
+    // Sets the leaf at the given index to the given hash.
+    // Panics if the index is out of bounds (since the structure doesn't grow).
+    pub fn set(&self, idx: usize, hash: Bytes32) {
+        #[cfg(feature = "counters")]
+        SET_COUNTERS[&self.ty].fetch_add(1, Ordering::Relaxed);
+        let mut layers = self.layers.lock();
+        if layers.data[0][idx] == hash {
             return;
         }
-        let mut next_hash = hash;
-        let empty_layers = &self.empty_layers;
-        let layers_len = self.layers.len();
-        for (layer_i, layer) in self.layers.iter_mut().enumerate() {
-            layer[idx] = next_hash;
-            if layer_i == layers_len - 1 {
-                // next_hash isn't needed
-                break;
-            }
-            let counterpart = layer
-                .get(idx ^ 1)
-                .cloned()
-                .unwrap_or_else(|| empty_layers[layer_i]);
-            if idx % 2 == 0 {
-                next_hash = hash_node(self.ty, next_hash, counterpart);
-            } else {
-                next_hash = hash_node(self.ty, counterpart, next_hash);
+        layers.data[0][idx] = hash;
+        layers.dirt[0].insert(idx >> 1);
+    }
+
+    /// Resizes the number of leaves the tree can hold.
+    ///
+    /// The extra space is filled with empty hashes.
+    pub fn resize(&self, new_len: usize) -> Result<usize, String> {
+        #[cfg(feature = "counters")]
+        RESIZE_COUNTERS[&self.ty].fetch_add(1, Ordering::Relaxed);
+        if new_len > self.capacity() {
+            return Err(
+                "Cannot resize to a length greater than the capacity of the tree.".to_owned(),
+            );
+        }
+        let mut layers = self.layers.lock();
+        let mut layer_size = new_len;
+        for (layer_i, layer) in layers.data.iter_mut().enumerate() {
+            layer.resize(layer_size, *empty_hash_at(self.ty, layer_i));
+            layer_size = max(layer_size >> 1, 1);
+        }
+        let start = layers.data[0].len();
+        for i in start..new_len {
+            layers.dirt[0].insert(i);
+        }
+        Ok(layers.data[0].len())
+    }
+}
+
+impl PartialEq for Merkle {
+    fn eq(&self, other: &Self) -> bool {
+        self.root() == other.root()
```

Comment:
Would it be possible for `min_depth` to differ here? Would such a distinction matter? I don't think it's possible but might be worth thinking through and adding a comment about.